rclcpp/src/rclcpp/client.cpp

```diff
@@ -42,3 +47,58 @@ ClientBase::get_client_handle() const
 {
   return &client_handle_;
 }
+
+bool
+ClientBase::service_is_ready() const
+{
+  bool is_ready;
+  rcl_ret_t ret =
+    rcl_service_server_is_available(this->get_rcl_node_handle(), &client_handle_, &is_ready);
+  if (ret != RCL_RET_OK) {
+    throw_from_rcl_error(ret, "rcl_service_server_is_available failed");
+  }
+  return is_ready;
+}
+
+bool
+ClientBase::wait_for_service_nanoseconds(std::chrono::nanoseconds timeout)
+{
+  auto start = std::chrono::steady_clock::now();
+  // check to see if the server is ready immediately
+  if (this->service_is_ready()) {
+    return true;
+  }
+  if (timeout == std::chrono::nanoseconds(0)) {
+    // check was non-blocking, return immediately
+    return false;
+  }
+  // make an event to reuse, rather than create a new one each time
+  auto event = node_->get_graph_event();
```

Comment:
Yes, I think so. I was avoiding this because basically we'll need to enable "make_shared_from_this" on Node. But I think that can't be avoided unless we rethink our ownership model a little bit.

I'll see what's involved to do that.